File: SingletonPattern/src/main/java/net/bi4vmr/study/singleton/java/TestSingleton.java
package net.bi4vmr.study.singleton.java;

/**
 * Name        : TestSingleton
 * <p>
 * Author      : BI4VMR
 * <p>
 * Email       : bi4vmr@outlook.com
 * <p>
 * Date        : 2023-09-29 21:44
 * <p>
 * Description : 测试类 - 单例模式。
 */
public class TestSingleton {

    public static void main(String[] args) {
        example05();
    }

    /*
     * 单例模式 - 饿汉式（非按需加载、线程安全）
     */
    static void example01() {
        // 调用SimpleSingleton的"getInstance"方法获取实例。
        SimpleSingleton instance = SimpleSingleton.getInstance();
        // 访问该对象的方法
        instance.function();
    }

    /*
     * 单例模式 - 懒汉式（按需加载、非线程安全）
     */
    static void example02() {
        // 使用多个线程获取实例
        for (int i = 1; i <= 3; i++) {
            final int index = i;
            new Thread(() -> {
                LazySingleton instance = LazySingleton.getInstance(0);
                System.out.println("线程" + index + "得到的对象ID：" + instance.getID());
            }).start();
        }
    }

    /*
     * 单例模式 - 懒汉式（按需加载、线程安全、"getInstance"方法效率不高）
     */
    static void example03() {
        // 使用多个线程获取实例
        for (int i = 1; i <= 3; i++) {
            final int index = i;
            new Thread(() -> {
                LazySyncSingleton instance = LazySyncSingleton.getInstance(0);
                System.out.println("线程" + index + "得到的对象ID：" + instance.getID());
            }).start();
        }
    }

    /*
     * 单例模式 - 懒汉式（按需加载、线程安全、"getInstance"方法效率较高）
     */
    static void example04() {
        // 使用多个线程获取实例
        for (int i = 1; i <= 3; i++) {
            final int index = i;
            new Thread(() -> {
                LazyDCLSingleton instance = LazyDCLSingleton.getInstance(0);
                System.out.println("线程" + index + "得到的对象ID：" + instance.getID());
            }).start();
        }
    }

    /*
     * 单例模式 - 懒汉式（按需加载、线程安全、实现简单、不支持初始化参数）
     */
    static void example05() {
        // 使用多个线程获取实例
        for (int i = 1; i <= 3; i++) {
            final int index = i;
            new Thread(() -> {
                LazyInnerClassSingleton instance = LazyInnerClassSingleton.getInstance();
                System.out.println("线程" + index + "得到的对象ID：" + instance.getID());
            }).start();
        }
    }
}

File: SingletonPattern/src/main/kotlin/net/bi4vmr/study/singleton/kt/TestSingleton.kt
package net.bi4vmr.study.singleton.kt

import kotlin.concurrent.thread

/**
 * Name        : TestSingleton
 *
 * Author      : BI4VMR
 *
 * Email       : bi4vmr@outlook.com
 *
 * Date        : 2023-10-01 13:38
 *
 * Description : 测试类 - 单例模式。
 */

fun main() {
    example06()
}

/*
 * 单例模式 - 饿汉式
 */
fun example01() {
    // 访问该对象的方法
    SimpleSingleton.function()
}

/*
 * 单例模式 - 懒汉式
 */
fun example02() {
    for (i in 1..3) {
        thread(start = true) {
            val instance: LazySingleton = LazySingleton.getInstance(0)
            println("线程" + i + "得到的对象ID：" + instance.id)
        }
    }
}

/*
 * 单例模式 - 懒汉同步式
 */
fun example03() {
    for (i in 1..3) {
        thread(start = true) {
            val instance: LazySyncSingleton = LazySyncSingleton.getInstance(0)
            println("线程" + i + "得到的对象ID：" + instance.id)
        }
    }
}

/*
 * 单例模式 - 双重校验锁式（不含初始化参数）
 */
fun example04() {
    for (i in 1..3) {
        thread(start = true) {
            val instance: LazyDCLSingleton = LazyDCLSingleton.getInstance()
            println("线程" + i + "得到的对象ID：" + instance.id)
        }
    }
}

/*
 * 单例模式 - 双重校验锁式（包含初始化参数）
 */
fun example05() {
    for (i in 1..3) {
        thread(start = true) {
            val instance: LazyDCLSingleton2 = LazyDCLSingleton2.getInstance(0)
            println("线程" + i + "得到的对象ID：" + instance.id)
        }
    }
}

/*
 * 单例模式 - 静态内部类式
 */
fun example06() {
    for (i in 1..3) {
        thread(start = true) {
            val instance: LazyInnerClassSingleton = LazyInnerClassSingleton.getInstance()
            println("线程" + i + "得到的对象ID：" + instance.id)
        }
    }
}
